### PR TITLE
BUG: name retention in Index.intersection

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

its going to take a few passes to get the behavior consistent across all subclasses.  once done, we can de-duplicate the boilerplate.

### CODE REVIEW

**Review by jreback — CHANGES_REQUESTED** [review]
likley will conflict with #36927 looks ok at a glance

**Comment by jreback** [comment]
pls add a whatsnew as well

**Comment by jbrockmendel** [comment]
whatsnew added + greenish (the same test_area_lim)

File: pandas/core/indexes/datetimelike.py [hunk 1]
```diff
@@ -686,10 +686,15 @@ def intersection(self, other, sort=False):
...
         self._validate_sort_keyword(sort)
         self._assert_can_do_setop(other)
+        other, _ = self._convert_can_do_setop(other)
 
         if self.equals(other):
             return self._get_reconciled_name_object(other)
 
+        return self._intersection(other, sort=sort)
+
+    def _intersection(self, other, sort=False):
```
Comment (jreback) [inline pandas/core/indexes/datetimelike.py:696]: type if you can (doc-string if needed)

File: pandas/core/indexes/interval.py [hunk 2]
```diff
@@ -124,7 +124,11 @@ def setop_check(method):
     def wrapped(self, other, sort=False):
         self._validate_sort_keyword(sort)
         self._assert_can_do_setop(other)
-        other = ensure_index(other)
+        other, _ = self._convert_can_do_setop(other)
+
+        if op_name == "intersection":
```
Comment (jreback) [inline pandas/core/indexes/interval.py:129]: why is this *only* the case?

**Review by jreback — CHANGES_REQUESTED** [review]
pls merge master

File: pandas/core/indexes/interval.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/indexes/interval.py:129]: bc there's a lot of cases to address and im handling them one at a time

**Review by jreback — APPROVED** [review]

**Pull request merged**